_posts/2019/2019-08-20-end-to-end-tests-with-geb-and-spock.md

```diff
@@ -0,0 +1,164 @@
+
+Be aware of the difference between repeating a test and retrying a test. When running stability tests, we do not want to retry any
+tests, because we do want to see the errors in this phase, so that we have a chance to fix them.
+
+
+## Concurrent test execution
+
+We had to take some precautions in order to be able to run tests concurrently.
+The word "concurrent" has two different meanings here. On one hand it refers to starting the program in a way that makes it invoke
+several tests in parallel. This can be achieved using the <a href="https://docs.gradle.org/current/dsl/org.gradle.api.tasks.testing.Test.html#org.gradle.api.tasks.testing.Test:maxParallelForks" target="_blank">`maxParallelForks`</a> property
+of a Gradle test task.
+This way of parallelising tests written in JVM / Gradle does not require any specific support by the test framework.
+
+The other meaning of "concurrent" refers to having two instances of the same test running *somewhere*.
+For example, a test is running in some CI pipeline, while a developer is starting the same test on their own computer.
+Why would this be a problem?
+Consider the case that some credentials are being used to log in to some third-party website or service. This could have the effect
+of invalidating any simultaneous session that might be going on, or otherwise interfering with the test result.
+
+Often times our third-party integrations would allow us to create any number of test accounts, but there could be exceptions.
+In one case we were not allowed to do that, as we were testing against the live system of that integration, because no "sandbox" existed.
+
+So we had to come up with a mechanism, that would allow us to prevent concurrent execution of this test, and instead wait for the
+ongoing run to finish. Conveniently, Java already provides <a href="https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/locks/Lock.html" target="_blank">an interface for this purpose</a>, which has exactly the desirable property that requesting a lock includes waiting for
+its availability, while disabling the current thread. Nice!  
+The interface is readily usable implemented by the <a href="https://github.com/redisson/redisson" target="_blank">Redisson</a> library,
+which uses <a href="https://redislabs.com/why-redis/" target="_blank">Redis</a> as its backend.
+
+
+## Diagnosing actual problems
+
+Of course, at some point, we want our end-to-end tests to be able to bring real problems with our software to our attention.
+Once random stability failures were mostly understood, this has been working well for us.
+
+The tests helped us to notice several issues related to third-party integrations and the checkout and order processes.
+
+In one case, we even found a critical issue, before it was reported to us via support tickets from our partners. You might be
+wondering, if the tests found the issue, why was it not fixed before it would get reported by live customers?
```

Comment:
This sentence reads strange?
It should be this: You may wonder, if the tests found the problem, why it was not fixed before it was reported by live customers?